Mathlib/Data/Matrix/Composition.lean

```diff
@@ -0,0 +1,132 @@
+
+/-- Mₙ(Mₘ(R)) ≃+* Mₙₘ(R) -/
+def comp_ringHom : Matrix I I (Matrix J J R) ≃+* Matrix (I × J) (I × J) R :=
+{ Matrix.comp_addHom I I J J R with
+  map_mul' := fun _ _ ↦ by
+    ext _ _
+    refine (Matrix.sum_apply _ _ _ _).trans $ Eq.symm Fintype.sum_prod_type
+}
+
+/-- Mₙ(Mₘ(R)) ≃+* Mₘ(Mₙ(R)) -/
+def swap_ringHom : Matrix (I × J) (I × J) R ≃+* Matrix (J × I) (J × I) R :=
+{ Matrix.swap_addHom I J I J R with
+  map_mul' := fun _ _ ↦ by
+    ext _ _
+    simp only [swap_addHom, swap, AddEquiv.toEquiv_eq_coe, Equiv.toFun_as_coe, EquivLike.coe_coe,
+      AddEquiv.coe_mk, Equiv.coe_fn_mk, mul_apply, Fintype.sum_prod_type]
+    exact Finset.sum_comm
+}
+
+end Semiring
+
+section Algebra
+
+variable (K : Type*) [CommSemiring K] [Semiring R] [Fintype I] [Fintype J] [Algebra K R]
+variable [DecidableEq I] [DecidableEq J]
+
+/-- Mₙ(Mₘ(R)) ≃ₐ[K] Mₙₘ(R) -/
+def comp_algHom : Matrix I I (Matrix J J R) ≃ₐ[K] Matrix (I × J) (I × J) R :=
+{ Matrix.comp_ringHom I J R with
+  commutes' := fun c ↦ by
+    ext ⟨i1, j1⟩ ⟨i2, j2⟩
+    simp only [comp_ringHom, comp_addHom, comp, AddEquiv.toEquiv_eq_coe, RingEquiv.toEquiv_eq_coe,
+      Equiv.toFun_as_coe, EquivLike.coe_coe, RingEquiv.coe_mk, AddEquiv.coe_mk, Equiv.coe_fn_mk]
+    simp only [algebraMap_eq_diagonal]; rw [Pi.algebraMap_def, Pi.algebraMap_def]
+    rw [@Algebra.algebraMap_eq_smul_one', @Algebra.algebraMap_eq_smul_one']
+    rw [← @diagonal_one, @diagonal_apply, @diagonal_apply]
+    if hii: i1 = i2 then
+      simp only [hii, ↓reduceIte, diagonal_one, smul_apply, Prod.mk.injEq, true_and]
+      if hjj : j1 = j2 then subst hjj; simp  else
+        simp only [ne_eq, hjj, not_false_eq_true, one_apply_ne, smul_zero, ↓reduceIte]
+    else simp only [hii, ↓reduceIte, zero_apply, Prod.mk.injEq]; tauto
+}
+
+/-- Mₙ(Mₘ(R)) ≃ₐ[K] Mₘ(Mₙ(R)) -/
+def swap_algHom : Matrix (I × J) (I × J) R ≃ₐ[K] Matrix (J × I) (J × I) R :=
+{ Matrix.swap_ringHom I J R with
+  commutes' := fun c ↦ by
+    ext ⟨i1, j1⟩ ⟨i2, j2⟩
+    simp only [swap_ringHom, swap_addHom, swap, AddEquiv.toEquiv_eq_coe, RingEquiv.toEquiv_eq_coe,
+      Equiv.toFun_as_coe, EquivLike.coe_coe, RingEquiv.coe_mk, AddEquiv.coe_mk, Equiv.coe_fn_mk,
+      algebraMap_eq_diagonal]; rw [Pi.algebraMap_def, Pi.algebraMap_def,
+      @Algebra.algebraMap_eq_smul_one', @diagonal_apply, @diagonal_apply]
+    if hii: i1 = i2 then
+      simp only [hii, ↓reduceIte, diagonal_one, smul_apply, Prod.mk.injEq, true_and]
+      if hjj : j1 = j2 then simp only [hjj, and_self, ↓reduceIte] else
+        simp only [ne_eq, hjj, not_false_eq_true, one_apply_ne, smul_zero, ↓reduceIte]; tauto
+    else simp only [Prod.mk.injEq, hii, and_false, ↓reduceIte, false_and]
+}
+
+
+open BigOperators Matrix MulOpposite in
+/-- Mn(Rᵒᵖ) ≃ₐ[K] Mₙ(R)ᵒᵖ -/
+def matrixEquivMatrixMop_algebra (n : ℕ):
+    Matrix (Fin n) (Fin n) Rᵐᵒᵖ ≃ₐ[K] (Matrix (Fin n) (Fin n) R)ᵐᵒᵖ where
+  toFun := fun M => MulOpposite.op (M.transpose.map (fun d => MulOpposite.unop d))
```

Comment:
You opened `MulOpposite` so probably `op` would work

Reply:
done